107-tee-boot-attestation.md

```diff
@@ -0,0 +1,498 @@
+this, initial vTPM state is tied to TEE attestation, and initial vTPM state will
+need to be available to run any workloads on CVMs. Thus, TEE attestation will be
+required for running any OS/applications and initializing initial vTPM state on
+a CVM.
+
+#### Story 1
+
+The use-case of this can be broken up into two main phases, being pre-boot TEE
+attestation and post-boot vTPM attestation.
+
+Pre-boot TEE attestation is performed from SVSM to unlock the CVM's root disk
+and establish initial vTPM state for the machine.
+
+Post-boot vTPM registration is performed by the Keylime agent and no functional
+changes are required. However, the agent enrollment policy will need to account
+for TEE attestation (as will be explained).
+
+Pre-boot Attestation
+--------------------
+
+1. Pre-boot VM Provisioning
+
+As mentioned before, CVM root disks are encrypted, with decryption reliant upon
+successful TEE attestation. In the VM provisioning phase, a user provisions an
+encrypted VM image sealed with initial vTPM state, which is in-turn encrypted
+with a key stored within Keylime. This step involves a new Keylime Registrar
+endpoint to fetch the public components of its "TEE Attestation Key" to then
+encrypt the initial vTPM state.
+
+- User creates an encrypted disk image instance from a plain text image
+  template.
+- User creates initial vTPM state for the CVM.
+- User seals encrypted disk image key against the initial vTPM state.
+- User makes POST request to new Keylime Registrar endpoint to fetch TEE
+  attestation key. Keylime Registrar replies with public components of TEE
+  attestation key.
+- User encrypts initial vTPM state with public TEE Attestation Key.
+- User stores encrypted initial vTPM state along with encrypted root disk.
+
+A diagram showing the VM provisioning process:
+```
+                                      ┌────────────────┐
+                                      │                │
+                                      │ ┌────────────┐ │
+                                      │ │VM rootfs   │ │
+                                      │ │Sealed w/   │ │
+┌──────┐                              │ │vTPM State  │ │
+│ VM   │    ┌──────────────┐          │ │Key         │ │
+│ Disk ├───►│ Provisioning ├─────────►│ └────────────┘ │
+│ Image│    │     Tool     │          │                │
+└──────┘    └───┬─────▲────┘          │ ┌────────────┐ │
+                │     │               │ │Initial     │ │
+         GET    │     │    TEE        │ │vTPM State  │ │
+         TEE    │     │Attestation    │ │encrypted w/│ │
+     Attestation│     │Public Key     │ │TEE         │ │
+     Public Key │     │               │ │Attestation │ │
+              ┌─▼─────┴─┐             │ │Key         │ │
+              │ Keylime │             │ └────────────┘ │
+              │Registrar│             │   Encrypted    │
+              └─────────┘             │     Image      │
+                                      └────────────────┘
```

Comment:
FYI, I'd expect this to be a multi phase process. Phase 1 creates an encrypted disk from a plain text template. Phase 2 manufactures new vTPM state and seals the disk against the vTPM.  Phase 3 encrypts the vTPM with a key held by the attestation server.

The reason for thinking of it as distinct phases is that there can be an arbitrary amount of time passing between phases 1 and 2 and 3. ie, don't assume they're all run at the same time in a single tool. They might be run together, or they might be run separately.

Reply:
Thanks. For time being I won't update the proposal to indicate this because from the server's POV it doesn't really matter. Keylime will add the endpoint to fetch the TEE Attestation Public Key and isn't concerned with what phase it's used, so long as the end result is a decryptable secret.